File: flink-demo/flinkJavaCode/src/main/java/com/yw/flink/example/javacases/case13_streamrelate/Case05_WindowCogroup.java
package com.yw.flink.example.javacases.case13_streamrelate;

import org.apache.flink.api.common.eventtime.SerializableTimestampAssigner;
import org.apache.flink.api.common.eventtime.WatermarkStrategy;
import org.apache.flink.api.common.functions.CoGroupFunction;
import org.apache.flink.api.java.functions.KeySelector;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.windowing.assigners.TumblingEventTimeWindows;
import org.apache.flink.streaming.api.windowing.time.Time;

import java.time.Duration;

/**
 * Flink - Window cogroup
 * 案例：读取订单流和支付流，进行window cogroup
 */
public class Case05_WindowCogroup {
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        //设置并行度为1
        env.setParallelism(1);

        //读取socket 订单流：订单ID，用户ID，订单金额，时间戳
        //order1,user1,10,1000
        //对订单流设置watermark
        SingleOutputStreamOperator<String> orderDSWithWatermark = env.socketTextStream("node5", 8888)
                .assignTimestampsAndWatermarks(
                        WatermarkStrategy.<String>forBoundedOutOfOrderness(Duration.ofSeconds(2))
                                .withTimestampAssigner((SerializableTimestampAssigner<String>) (s, l) -> Long.parseLong(s.split(",")[3]))
                );

        //读取socket 支付流：订单ID，支付金额，支付时间戳
        //order1,10,1000
        //对支付流设置watermark
        SingleOutputStreamOperator<String> payDSWithWatermark = env.socketTextStream("node5", 9999)
                .assignTimestampsAndWatermarks(
                        WatermarkStrategy.<String>forBoundedOutOfOrderness(Duration.ofSeconds(2))
                                .withTimestampAssigner((SerializableTimestampAssigner<String>) (s, l) -> Long.parseLong(s.split(",")[2]))
                );

        //对两个流进行window cogroup
        orderDSWithWatermark.coGroup(payDSWithWatermark)
                .where((KeySelector<String, String>) s -> s.split(",")[0])
                .equalTo((KeySelector<String, String>) s -> s.split(",")[0])
                .window(TumblingEventTimeWindows.of(Time.seconds(5)))
                .apply((CoGroupFunction<String, String, String>) (left, right, collector) -> collector.collect(left + " ==== " + right)).print();

        env.execute();
    }
}

File: flink-demo/flinkScalaCode/src/main/scala/com/yw/flink/example/scalacases/case13_streamrelate/Case05_WindowCogroup.scala
package com.yw.flink.example.scalacases.case13_streamrelate

import org.apache.flink.api.common.eventtime.{SerializableTimestampAssigner, WatermarkStrategy}
import org.apache.flink.api.common.functions.CoGroupFunction
import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment
import org.apache.flink.streaming.api.windowing.assigners.TumblingEventTimeWindows
import org.apache.flink.streaming.api.windowing.time.Time
import org.apache.flink.util.Collector

import java.lang
import java.time.Duration

/**
 * Flink - Window Join
 * 案例：读取订单流和支付流，进行window cogroup
 */
object Case05_WindowCogroup {
  def main(args: Array[String]): Unit = {
    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
    env.setParallelism(1)
    //导入隐式转换
    import org.apache.flink.streaming.api.scala._

    //读取socket 订单流：订单ID，用户ID，订单金额，时间戳
    //order1,user1,10,1000
    val orderDS: DataStream[String] = env.socketTextStream("node5", 8888)
    //设置订单流的watermark
    val orderDSWithWatermark: DataStream[String] = orderDS.assignTimestampsAndWatermarks(
      WatermarkStrategy.forBoundedOutOfOrderness[String](Duration.ofSeconds(2))
        .withTimestampAssigner(new SerializableTimestampAssigner[String] {
          override def extractTimestamp(t: String, l: Long): Long = t.split(",")(3).toLong
        })
    )

    //读取socket 支付流：订单ID，支付金额，支付时间戳
    //order1,10,1000
    val payDS: DataStream[String] = env.socketTextStream("node5", 9999)
    //设置支付流的watermark
    val payDSWithWatermark: DataStream[String] = payDS.assignTimestampsAndWatermarks(
      WatermarkStrategy.forBoundedOutOfOrderness[String](Duration.ofSeconds(2))
        .withTimestampAssigner(new SerializableTimestampAssigner[String] {
          override def extractTimestamp(t: String, l: Long): Long = t.split(",")(2).toLong
        })
    )

    //对两个流进行window cogroup
    orderDSWithWatermark.coGroup(payDSWithWatermark)
      .where(_.split(",")(0))
      .equalTo(_.split(",")(0))
      .window(TumblingEventTimeWindows.of(Time.seconds(5)))
      .apply(new CoGroupFunction[String,String,String] {
        override def coGroup(left: lang.Iterable[String],
                             right: lang.Iterable[String],
                             collector: Collector[String]): Unit = {
          collector.collect(left+" ==== "+right)
        }
      }).print()
    env.execute()

  }

}